src/functions/boundaryAreaOverlap.ts

```diff
@@ -1,96 +1,110 @@
-        if (
-          !isInternalVectorDatasource(ds) &&
-          !isExternalVectorDatasource(ds)
-        ) {
-          throw new Error(`Expected vector datasource for ${ds.datasourceId}`);
-        }
+  const clippedSketch = await clipSketchToSubregion(sketch, extraParams!);
 
-        // Fetch only the features that overlap the bounding box of the sketch
-        const url = project.getVectorDatasourceUrl(ds);
-        const polys = await getFeatures(ds, url, {
-          bbox: sketchBox,
-        });
-        if (!isPolygonFeatureArray(polys)) {
-          throw new Error("Expected array of Polygon features");
-        }
-        return polys;
-      })
-    )
-  ).reduce<Record<string, Feature<Polygon>[]>>((acc, polys, classIndex) => {
-    return {
-      ...acc,
-      [metricGroup.classes[classIndex].classId]: polys,
-    };
-  }, {});
+  const areaMetrics = (
+    await overlapArea(metricGroup.metricId, sketch, totalAreaMetric.value, {
+      includePercMetric: false,
+    })
+  ).map(
+    (metrics): Metric => ({
+      ...metrics,
+      classId: metricGroup.classes[0].classId,
+    })
+  );
 
-  const metrics: Metric[] = ( // calculate area overlap metrics for each class
-    await Promise.all(
-      metricGroup.classes.map(async (curClass) => {
-        const overlapResult = await overlapFeatures(
-          metricGroup.metricId,
-          polysByBoundary[curClass.classId],
-          sketch
-        );
-        return overlapResult.map(
-          (metric): Metric => ({
-            ...metric,
-            classId: curClass.classId,
-          })
-        );
-      })
-    )
-  ).reduce(
-    // merge
-    (metricsSoFar, curClassMetrics) => [...metricsSoFar, ...curClassMetrics],
-    []
+  // Generate area metrics grouped by protection level, with area overlap within protection level removed
+  // Each sketch gets one group metric for its protection level, while collection generates one for each protection level
+  const sketchToMpaClass = getSketchToMpaProtectionLevel(
+    clippedSketch,
+    areaMetrics
   );
-  console.log("the  metrics", metrics);
+  const metricToLevel = (sketchMetric: Metric) => {
+    return sketchToMpaClass[sketchMetric.sketchId!];
+  };
+
+  const levelMetrics = await overlapAreaGroupMetrics({
+    metricId: metricGroup.metricId,
+    groupIds: ["FULLY_PROTECTED", "HIGHLY_PROTECTED"],
+    sketch: clippedSketch,
+    metricToGroup: metricToLevel,
+    metrics: areaMetrics,
+    classId: metricGroup.classes[0].classId,
+    outerArea: totalAreaMetric.value,
+  });
 
   return {
-    metrics: sortMetrics(rekeyMetrics(metrics)),
-    sketch: toNullSketch(sketch, true),
+    metrics: sortMetrics(rekeyMetrics([...areaMetrics, ...levelMetrics])),
+    sketch: toNullSketch(clippedSketch),
   };
 }
 
+export function getSketchToMpaProtectionLevel(
+  sketch: Sketch | SketchCollection | NullSketchCollection | NullSketch,
+  metrics: Metric[]
```

Comment:
Add code comment?  Consider just `sketchToMpaProtectionLevel` or `getMpaProtectionLevel`.  The combo isn't clear to me.

Reply:
Agreed, I prefer `getMpaProtectionLevel`, I'll switch it over and add a function comment.